### ISSUE TITLE
[scm] Expressions not expanded when "auto" mode

### ISSUE THREAD

**Issue by niosHD** [conan-io/conan#3183]
Conan Version: 1.5.2

I just played around some more with the scm feature and found a (for me as user) surprising side effect. As soon as either the "url" or the "revision" field is in "auto" mode, not only the respective field is rewritten but also other fields in the scm structure are rewritten to their current value. For standard strings it does not matter, however, as soon as a function is called to provide the value, this becomes visible.

In more detail, exporting a recipe with the following scm structure yields the expected result and exactly the same structure is found in the conan cache.

    scm = {
        "type": "git",
        "subfolder": "repo",
        "username": tools.get_env("MY_MAGIC_USERNAME", "error"),
        "password": tools.get_env("MY_MAGIC_PASSWORD", "error"),
        "url": "http://example.com",
        "revision": "1234"
    }

However, as soon as I set, for example, the revision to "auto", all fields are evaluated and exported as their actual value at export time. This can then look something like the following:

    scm = {"password": "error",
           "revision": "1e25d838c88afaa6f190d12ea9dff9de36450f0c",
           "subfolder": "repo",
           "type": "git",
           "url": "http://example.com",
           "username": "error"}

What bothers me is that the behaviour feels a little bit inconsistent which makes me wonder if this is a bug or a feature?

On the one hand, I dislike that it happens as side effect of capturing the url/revision. On the other hand, I like the possibility to capture values at export time into the recipe. Having a generic measure like this would, for example, permit to easily implement #3150 inside the recipe without further conan support. I imagine something like an `expand` function, that can be used to initialize arbitrary properties, and is replaced with its result when the recipe is exported. Implementing #3150 would then look something like the following where `get_revision_from_branch` is defined as free function in the recipe.

    scm = {
        "type": "git",
        "url": "http://example.com",
        "revision": expand(get_revision_from_branch("feature/foo"))
    }

Similarly, #3160 could be extended to the following in order to remove the `get_version` call when consuming the recipe.

    version = expand(get_version())

What do you guys think about the idea? Is it worthwhile pursuing and can it be implemented at all?

**Comment by lasote** [comment]
Hi!

The problem is if you specify the following `scm` code in a `conanfile.py`:

```
scm = {
    "type": "git",
    "subfolder": "repo",
    "username": tools.get_env("MY_MAGIC_USERNAME", "error"),
    "password": tools.get_env("MY_MAGIC_PASSWORD", "error"),
    "url": "http://example.com",
    "revision": "1234"
}
```

the `tools.get_env` is evaluated when loading the class, there is nothing conan could do with it. Is python stuff.

The only way we could try to operate is if you declare a lambda function instead of a value, we could (theoretically speaking) evaluate the lambda only when accessing the field in the scm. So the only way I see to fix this is the opposite to your proposal, by default everything is expanded, but we could evaluate later a function.

**Comment by niosHD** [comment]
I would also be fine with having expansion by default as long as it is consistent and there is a way to opt-out. If this is possible by specifying a lambda then this is fine by me.

The reason why I proposed an `expand` function as opt-in is the following. Without actually looking at conan's implementation for scm feature, my idea was to wrap the initialization value using something like an `ExpandValue` class via the `expand` function. Subsequently, the same logic, that is currently used to rewrite the scm block when url/revision is specified as "auto", could be used on every class member guarded by an `isInstance(member, ExpandValue)` check.

**Comment by lasote** [comment]
To summarize how the feature could look:

- The `scm` attribute in the conanfile could have some attributes with "expressions" or "lambdas" or something we don't have to evaluate at the `get_scm_data`. They should be treated as expressions (as they were strings => Note that a `lambda` as is won't work, because we don't want a `lambda object` we want the `get_revision_from_branch("feature/foo")` as a literal expression`
- So when the values are replaced, the data to be replaced is not the evaluation is the literal expression.
- To help understand it, an ugly example implementation could be:

```
scm = {
    "type": "git",
    "subfolder": "repo",
    "username": "#EXPRESSION#tools.get_env('MY_MAGIC_USERNAME', 'error')",
    "password":  "#EXPRESSION#tools.get_env('MY_MAGIC_PASSWORD', 'error')",
    "url": "http://example.com",
    "revision": "1234"
}
``` 

And in the replacement:
```
if attribute is string and starts with "#EXPRESSION#":
    attribute_str = value.split("#EXPRESSION#")[1]
```

**Comment by lasote** [comment]
So, @niosHD  in addition to "username" and "password" do you have any other use case for this "not expanded variables"? I can't imagine any other, maybe the feature could be simplified as declaring standard `CONAN_SCM_USERNAME` and `CONAN_SCM_PASSWORD` to be used when no username and password is specified?

**Comment by niosHD** [comment]
Theoretically, the same problem exists for each field that exists within the `scm` property. For example, if someone wants to write a recipe where a command line variable permits to select if either an ssh or an https clone url is used, doing that is not straight forward possible using only the `scm` feature. Instead, I assume that a custom `source` method has to be used in addition to the `scm` auto deduction.

Still, to be perfectly honest and assuming that it works, I am completely fine with writing a custom `source` method in such scenarios. I just brought up the topic because the `scm` feature did not behave like I expected it to. Furthermore, I feel strongly that the whole `scm` auto deduction functionality is only a special case of a more generic and more useful feature (see #3338).

In summary, adding the eval like behavior with code in strings is in my opinion not very nice and most likely not worth the effort. It is probably best to document the whole situation as a limitation of the current `scm` feature and move on. Hopefully, a more generic way to capture and store information at export time makes the whole problem obsolete in the future.

Anyway, thank you @lasote for looking into this issue! I really appreciate the effort you guys put into conan and how you interact with the community.

**Comment by lasote** [comment]
Hi again! Please, take a look at #4088, I think it solves your issue.
